### PR TITLE
Implement DNS server

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: networkmanager/dnsserver.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		binary:         dnsMasqBinary,
+		hosts:          make(map[string][]string),
+	}
+
+	if err := dnsServer.prepareDNSConfFile(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if err := dnsServer.restart(); err != nil {
+		return nil, aoserrors.Wrap(err)
```
Comment (al1img) [inline networkmanager/dnsserver.go:117]: ditto

File: networkmanager/networkmanager.go [hunk 2]
```diff
@@ -79,14 +81,19 @@ var (
  **********************************************************************************************************************/
 
 // New creates network manager instance.
-func New(storage Storage) (*NetworkManager, error) {
+func New(storage Storage, workingDir string) (*NetworkManager, error) {
```
Comment (al1img) [inline networkmanager/networkmanager.go:84]: path whole config

File: networkmanager/dnsserver.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+	output, err := ExecContext(dns.binary, args...).CombinedOutput()
+	if err != nil {
+		return aoserrors.Errorf("message: %s, err: %v", string(output), err)
+	}
+
+	return nil
+}
+
+func (dns *dnsServer) restartProcess(pid *os.Process) error {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/dnsserver.go:270]: Done

File: networkmanager/dnsserver.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+		configFile:     filepath.Join(networkDir, confFileName),
+		PidFile:        filepath.Join(networkDir, pidFileName),
+		AddOnHostsFile: filepath.Join(networkDir, hostsFileName),
+		IPAddress:      ip.String(),
+		binary:         dnsMasqBinary,
+		hosts:          make(map[string][]string),
+	}
+
+	if err := dnsServer.prepareDNSConfFile(); err != nil {
+		return nil, aoserrors.Wrap(err)
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/dnsserver.go:113]: Done

File: networkmanager/dnsserver.go (hunk 1 above)
Comment (MykolaSolyanko) [reply networkmanager/dnsserver.go:117]: Done

File: networkmanager/dnsserver.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+	}
+
+	if err := dnsServer.restart(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	return dnsServer, nil
+}
+
+func (dns *dnsServer) getAddress() string {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/dnsserver.go:123]: Done

File: networkmanager/dnsserver.go [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+func (dns *dnsServer) rewriteHostsFile() error {
+	f, err := os.OpenFile(dns.AddOnHostsFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Errorf("failed to close hosts file: %v", err)
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/dnsserver.go:155]: Done

File: launcher/launcher.go [hunk 7]
```diff
@@ -751,8 +756,21 @@ func (launcher *Launcher) prepareInstanceStartInfo(service imagemanager.ServiceI
 		return instanceInfo, aoserrors.Errorf("can't setup storage and state for instance: %v", err)
 	}
 
+	var hosts []string
+	if service.Config.Hostname != nil {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply launcher/launcher.go:760]: Done

File: networkmanager/networkmanager.go (hunk 2 above)
Comment (MykolaSolyanko) [reply networkmanager/networkmanager.go:84]: Done

File: networkmanager/dnsserver.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+// These global variable is used to be able to mocking the functionality of networking in tests.
+// nolint:gochecknoglobals
+var (
+	LookPath          = exec.LookPath
+	DiscoverInterface = gateway.DiscoverInterface
+	ExecContext       = newExecShellCommander
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/dnsserver.go:85]: Done

File: networkmanager/dnsserver.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+expand-hosts
+pid-file={{.PidFile}}
+except-interface=lo
+bind-dynamic
+no-hosts
+listen-address={{.IPAddress}}
+addn-hosts={{.AddOnHostsFile}}`
+)
+
+type ShellCommand interface {
```
Comment (al1img) [inline networkmanager/dnsserver.go:60]: Unused interface.

File: .golangci.yml [hunk 10]
```diff
@@ -48,7 +48,7 @@ linters-settings:
   gomnd:
     settings:
       mnd:
-        ignored-numbers: 0o600,0o755,5,10,16,32,64,128,100,0xff,3
+        ignored-numbers: 0o600,0o755,0o644,5,10,16,32,64,128,100,0xff,3
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply .golangci.yml:51]: Done

File: launcher/launcher.go [hunk 11]
```diff
@@ -751,8 +756,22 @@ func (launcher *Launcher) prepareInstanceStartInfo(service imagemanager.ServiceI
 		return instanceInfo, aoserrors.Errorf("can't setup storage and state for instance: %v", err)
 	}
 
+	var hosts []string
+
+	if service.Config.Hostname != nil {
+		hosts = append(hosts, *service.Config.Hostname)
+	}
+
+	if instance.ServiceID != "" && instance.SubjectID != "" {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply launcher/launcher.go]: Done

File: networkmanager/dnsserver.go (hunk 9 above)
Comment (MykolaSolyanko) [reply networkmanager/dnsserver.go]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**